### PR TITLE
Add supported package definition type

### PR DESCRIPTION
**Pull request by larioj**
We are adding a new type to cosmos. This type will represent the supported packaging formats. The  idea is that we will use this type in contexts where we only want to support certain packages.

### CODE REVIEW

File: cosmos-model/src/main/scala/com/mesosphere/universe/v3/model/PackageDefinition.scala [hunk 1]
```diff
@@ -128,11 +129,100 @@ case class V3Package(
...
+) extends PackageDefinition with Ordered[V3Package] with Spd {
   override def compare(that: V3Package): Int = {
     PackageDefinition.compare(
       (name, version, releaseVersion),
       (that.name, that.version, that.releaseVersion)
     )
   }
 }
+
+sealed trait Spd
```
Comment (cruhland) [inline cosmos-model/src/main/scala/com/mesosphere/universe/v3/model/PackageDefinition.scala]: Please rename this to something understandable to someone unfamiliar with the intent of the code.

Also, you might want to use a "self type" to restrict what this can be mixed into:

```scala
sealed trait Spd { self: PackageDefinition => }
```

File: cosmos-model/src/main/scala/com/mesosphere/universe/v3/model/PackageDefinition.scala [hunk 2]
```diff
@@ -128,11 +129,100 @@ case class V3Package(
...
+
+object Spd {
+  implicit class SpdOps(spd: Spd) {
```
Comment (cruhland) [inline cosmos-model/src/main/scala/com/mesosphere/universe/v3/model/PackageDefinition.scala]: Rename this too.

Comment (cruhland) [inline cosmos-model/src/main/scala/com/mesosphere/universe/v3/model/PackageDefinition.scala]: For full efficiency, this should really be:

```scala
implicit final class SpdOps(val spd: Spd) extends AnyVal {
```

Extending `AnyVal` means that there won't be an instance of `SpdOps` at runtime -- all members are treated as having Java's `static` modifier, essentially.

Comment (cruhland) [inline cosmos-model/src/main/scala/com/mesosphere/universe/v3/model/PackageDefinition.scala]: Is there a base trait that defines some of these ops already for `PackageDefinition`? I think you might be able to mix that trait into this class, to ensure we define all the expected methods.

**Comment by cruhland** [comment]
@larioj left a few comments.

File: cosmos-http/src/main/scala/com/mesosphere/universe/MediaTypes.scala [hunk 3]
```diff
@@ -31,4 +31,10 @@ object MediaTypes {
     Map("version" -> "v1")
   )
 
+  val supportedPackageDefinition = MediaType(
+    "application",
+    MediaTypeSubType("vnd.dcos.universe.supported-package", Some("json")), // TODO: @larioj check this
```
Comment (cruhland) [inline cosmos-http/src/main/scala/com/mesosphere/universe/MediaTypes.scala]: This shouldn't exist; the package and/or repo objects we send and receive should always be of a concrete type. Unless I'm misunderstanding what this is for?

Comment (larioj) [inline cosmos-http/src/main/scala/com/mesosphere/universe/MediaTypes.scala]: This is for storage. The ObjectStorage API requires a media type. But since this will be S3, then yes, I should replace this.

Comment (larioj) [inline cosmos-http/src/main/scala/com/mesosphere/universe/MediaTypes.scala]: At the moment we don't use the media type for anything.

File: cosmos-json/src/main/scala/com/mesosphere/universe/v3/circe/Decoders.scala [hunk 4]
```diff
@@ -118,4 +118,15 @@ object Decoders {
...
+    Decoder.instance[SupportedPackageDefinition] { (hc: HCursor) =>
+      hc.downField("packagingVersion").as[PackagingVersion].flatMap {
+        case V3PackagingVersion => hc.as[V3Package]
+        case V2PackagingVersion => Left(DecodingFailure(
+          s"V2Package is not a supported package definition",
+          hc.history
+        ))
+      }
+    }
+  }
```
Comment (jsancio) [inline cosmos-json/src/main/scala/com/mesosphere/universe/v3/circe/Decoders.scala]: Where do we use this decoder?

File: cosmos-model/src/main/scala/com/mesosphere/universe/v3/model/PackageDefinition.scala [hunk 5]
```diff
@@ -128,11 +128,28 @@ case class V3Package(
...
   override def compare(that: V3Package): Int = {
     PackageDefinition.compare(
       (name, version, releaseVersion),
       (that.name, that.version, that.releaseVersion)
     )
   }
 }
+
+sealed trait SupportedPackageDefinition extends PackageDefinition {
+  self: PackageDefinition =>
```
Comment (jsancio) [inline cosmos-model/src/main/scala/com/mesosphere/universe/v3/model/PackageDefinition.scala]: Do you need `self: PackageDefinition =>` given that this trait extends `PackageDefinition`?

File: cosmos-model/src/main/scala/com/mesosphere/universe/v3/model/PackageDefinition.scala [hunk 6]
```diff
@@ -128,11 +128,28 @@ case class V3Package(
   resource: Option[V3Resource] = None,
   config: Option[JsonObject] = None,
   command: Option[Command] = None
-) extends PackageDefinition with Ordered[V3Package] {
+) extends PackageDefinition with Ordered[V3Package] with SupportedPackageDefinition {
```
Comment (jsancio) [inline cosmos-model/src/main/scala/com/mesosphere/universe/v3/model/PackageDefinition.scala]: Change this to `extends SupportedPackageDefinition with Ordered[V3Package]` since `S..P..D` extends `PackageDefinition`.

File: cosmos-model/src/main/scala/com/mesosphere/universe/v3/syntax/PackageDefinitionOps.scala [hunk 7]
```diff
@@ -134,7 +134,9 @@ final class PackageDefinitionOps(val pkgDef: PackageDefinition) extends AnyVal {
 object PackageDefinitionOps {
   import scala.language.implicitConversions
 
-  implicit def packageDefinitionToPackageDefinitionOps(pkgDef: PackageDefinition): PackageDefinitionOps = {
+  implicit def packageDefinitionToPackageDefinitionOps[P <: PackageDefinition](
+    pkgDef: P
+  ): PackageDefinitionOps = {
```
Comment (jsancio) [inline cosmos-model/src/main/scala/com/mesosphere/universe/v3/syntax/PackageDefinitionOps.scala]: Why do we need a type parameter? What error do you get with the old signature?

Comment (larioj) [inline cosmos-model/src/main/scala/com/mesosphere/universe/v3/syntax/PackageDefinitionOps.scala]: The type parameter makes it so that this conversion works for SPD, since it is a subtype of PackageDefinition. Without the type parameter the implicit conversion does not get picked up.

File: cosmos-json/src/main/scala/com/mesosphere/universe/v3/circe/Decoders.scala (hunk 4 above)
Comment (larioj) [inline cosmos-json/src/main/scala/com/mesosphere/universe/v3/circe/Decoders.scala]: We need this decoder for all the other decoders that depend on it. Since we are changing the Operation class to include SPD. The Operation class is used in a bunch of other classes that have decoders.

File: cosmos-model/src/main/scala/com/mesosphere/universe/v3/model/PackageDefinition.scala (hunk 5 above)
Comment (larioj) [inline cosmos-model/src/main/scala/com/mesosphere/universe/v3/model/PackageDefinition.scala]: I'll remove it.

File: cosmos-model/src/main/scala/com/mesosphere/universe/v3/model/PackageDefinition.scala (hunk 6 above)
Comment (larioj) [inline cosmos-model/src/main/scala/com/mesosphere/universe/v3/model/PackageDefinition.scala]: Done

File: cosmos-model/src/main/scala/com/mesosphere/universe/v3/syntax/PackageDefinitionOps.scala (hunk 7 above)
Comment (jsancio) [inline cosmos-model/src/main/scala/com/mesosphere/universe/v3/syntax/PackageDefinitionOps.scala]: Okay.

File: cosmos-model/src/main/scala/com/mesosphere/cosmos/rpc/v1/model/AddResponse.scala [hunk 8]
```diff
@@ -4,4 +4,4 @@ import com.mesosphere.universe
 
 // This class is only a wrapper for the underlying package
 // We need a unique type so that the media type encoder can be resolved implicitly
-final class AddResponse(val v3Package: universe.v3.model.V3Package) extends AnyVal
+final class AddResponse(val supportedPackage: universe.v3.model.SupportedPackageDefinition) extends AnyVal
```
Comment (cruhland) [inline cosmos-model/src/main/scala/com/mesosphere/cosmos/rpc/v1/model/AddResponse.scala]: I'm wondering if this is applying abstraction in the wrong place. This file defines what an `rpc.v1.model.AddResponse` is, which is a `V3Package`. If we want to have an `AddResponse` as a `V4Package`, we should create an `rpc.v2.model.AddResponse`. The decision about which version to send would then be made by an instance of `DispatchingMediaTypedEncoder[rpc.v2.model.AddResponse]`. See [here](https://github.com/dcos/cosmos/blob/master/cosmos-server/src/main/scala/com/mesosphere/cosmos/rpc/v2/circe/MediaTypedEncoders.scala#L30) for an example.

File: cosmos-model/src/main/scala/com/mesosphere/universe/v3/model/PackageDefinition.scala [hunk 9]
```diff
@@ -128,11 +128,26 @@ case class V3Package(
...
       (that.name, that.version, that.releaseVersion)
     )
   }
 }
+
+sealed trait SupportedPackageDefinition extends PackageDefinition
+
+object SupportedPackageDefinition {
+
+  implicit val supportedPackageDefinitionOrdering = new Ordering[SupportedPackageDefinition] {
```
Comment (cruhland) [inline cosmos-model/src/main/scala/com/mesosphere/universe/v3/model/PackageDefinition.scala]: Why is this needed now?

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/handler/PackageAddHandler.scala [hunk 10]
```diff
@@ -62,9 +62,9 @@ final class PackageAddHandler(
 
     for {
       operation <- futureOperation
-      _ <- producerView.add(operation.v3Package.packageCoordinate, operation)
+      _ <- producerView.add(operation.packageDefinition.packageCoordinate, operation)
     } yield {
-      new rpc.v1.model.AddResponse(operation.v3Package)
+      new rpc.v1.model.AddResponse(operation.packageDefinition)
```
Comment (cruhland) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/handler/PackageAddHandler.scala]: Ah, this is why you changed the definition of `AddResponse`. We need to be careful about changing the definitions of anything under a `model` package: they are intended to be immutable once defined (except for changes which would not affect clients). Instead we use the version numbering (`v1.model`, `v2.model`, etc.) to account for changes. This ensures that we do not break old clients when releasing a new version of Cosmos.

File: cosmos-model/src/main/scala/com/mesosphere/cosmos/rpc/v1/model/AddResponse.scala (hunk 8 above)
Comment (jsancio) [inline cosmos-model/src/main/scala/com/mesosphere/cosmos/rpc/v1/model/AddResponse.scala]: I think we need to add `universe.v5.model.SupportPackageDefinition` if we add support for v5 packages.

Comment (cruhland) [inline cosmos-model/src/main/scala/com/mesosphere/cosmos/rpc/v1/model/AddResponse.scala]: Ah, if we're going to have a new version of `SupportedPackageDefinition` each time we add a new package version, then this change should be fine. But doesn't that kind of defeat the point of having the `SupportedPackageDefinition` trait in the first place? Since we'd have to update a bunch of places in the code to use the new version? Or is the idea that we'd keep all the existing code as is, and only new code would use the new version?

Comment (cruhland) [inline cosmos-model/src/main/scala/com/mesosphere/cosmos/rpc/v1/model/AddResponse.scala]: For example, what about low-level, generic code that doesn't care which version of a package it's handling?

Comment (larioj) [inline cosmos-model/src/main/scala/com/mesosphere/cosmos/rpc/v1/model/AddResponse.scala]: Does this mean that we would have to duplicate endpoints?

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/handler/PackageAddHandler.scala (hunk 10 above)
Comment (larioj) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/handler/PackageAddHandler.scala]: Oh yeah, I thought we could break the experimental stuff. Or should I not assume that?

**Comment by cruhland** [comment]
@larioj looks good, I'm just a bit uncertain about some use cases; where should or shouldn't we be using `SupportedPackageDefinition`?

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/handler/PackageAddHandler.scala (hunk 10 above)
Comment (jsancio) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/handler/PackageAddHandler.scala]: You can break the experimental stuff but let's make sure that we update the CLI if it breaks the CLI.

**Comment by larioj** [comment]
# Notes on the latest changes

* Add response contains a v3Package.
* Package storage decodes types based on the media type. This is not compiler checked.
* The commits are actually well organized.

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/storage/PackageStorage.scala [hunk 11]
```diff
@@ -37,8 +38,13 @@ final class PackageStorage private(objectStorage: ObjectStorage) {
   ): Future[Option[universe.v3.model.SupportedPackageDefinition]] = {
     val metadataName = getMetadataPath(packageCoordinate)
     objectStorage.readAsArray(metadataName).map { pendingRead =>
-      pendingRead.map { case (_, data) =>
-        decode[universe.v3.model.SupportedPackageDefinition](new String(data, StandardCharsets.UTF_8))
+      pendingRead.map {
+        case (universe.MediaTypes.universeV3Package, data) =>
```
Comment (larioj) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/storage/PackageStorage.scala]: This case statement is not checked by the compiler.

File: cosmos-bijection/src/main/scala/com/mesosphere/universe/bijection/UniverseConversions.scala [hunk 12]
```diff
@@ -124,6 +126,15 @@ object UniverseConversions {
     )
   }
 
+  implicit val MetadataToSupportedPackageDefinition:
```
Comment (cruhland) [inline cosmos-bijection/src/main/scala/com/mesosphere/universe/bijection/UniverseConversions.scala]: Nitpick: this name shouldn't be capitalized

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/storage/PackageStorage.scala [hunk 13]
```diff
@@ -37,8 +38,13 @@ final class PackageStorage private(objectStorage: ObjectStorage) {
...
+          decode[universe.v3.model.SupportedPackageDefinition](
```
Comment (larioj) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/storage/PackageStorage.scala]: maybe I should make this `decode[universe.v3.model.V3Package]`

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/handler/PackageAddHandler.scala [hunk 14]
```diff
@@ -65,7 +65,10 @@ final class PackageAddHandler(
       operation <- futureOperation
       _ <- producerView.add(operation.packageDefinition.packageCoordinate, operation)
     } yield {
-      new rpc.v1.model.AddResponse(operation.packageDefinition)
+      operation.packageDefinition match {
+        case v3Package: universe.v3.model.V3Package =>
```
Comment (larioj) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/handler/PackageAddHandler.scala]: When we add a v4 package, we will have to extend this case statement. I'm not sure, what I need to do when this is V4Package.

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/storage/PackageStorage.scala [hunk 15]
```diff
@@ -22,21 +23,28 @@ final class PackageStorage private(objectStorage: ObjectStorage) {
...
   def readPackageDefinition(
     packageCoordinate: rpc.v1.model.PackageCoordinate
-  ): Future[Option[universe.v3.model.V3Package]] = {
+  ): Future[Option[universe.v3.model.SupportedPackageDefinition]] = {
     val metadataName = getMetadataPath(packageCoordinate)
     objectStorage.readAsArray(metadataName).map { pendingRead =>
-      pendingRead.map { case (_, data) =>
-        decode[universe.v3.model.V3Package](new String(data, StandardCharsets.UTF_8))
+      pendingRead.map {
+        case (universe.MediaTypes.universeV3Package, data) =>
```
Comment (larioj) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/storage/PackageStorage.scala]: This case statement is not checked by the compiler.

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/handler/PackageAddHandler.scala [hunk 16]
```diff
@@ -55,23 +55,26 @@ final class PackageAddHandler(
...
     }
 
     for {
       operation <- futureOperation
-      _ <- producerView.add(operation.v3Package.packageCoordinate, operation)
+      _ <- producerView.add(operation.packageDefinition.packageCoordinate, operation)
     } yield {
-      new rpc.v1.model.AddResponse(operation.v3Package)
+      operation.packageDefinition match {
+        case v3Package: universe.v3.model.V3Package =>
```
Comment (larioj) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/handler/PackageAddHandler.scala]: This case statement will eventually include a v4 package. I'm not sure what the match will return.

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/storage/PackageStorage.scala [hunk 17]
```diff
@@ -22,21 +23,28 @@ final class PackageStorage private(objectStorage: ObjectStorage) {
...
+          decode[universe.v3.model.SupportedPackageDefinition](
```
Comment (larioj) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/storage/PackageStorage.scala]: maybe I should make this `decode[universe.v3.model.V3Package]`

File: cosmos-bijection/src/main/scala/com/mesosphere/universe/bijection/UniverseConversions.scala [hunk 18]
```diff
@@ -1,6 +1,8 @@
 package com.mesosphere.universe.bijection
 
 import com.mesosphere.universe
+import com.mesosphere.universe.v3.model.V3Metadata
+import com.mesosphere.universe.v3.model.V3Package
```
Comment (larioj) [inline cosmos-bijection/src/main/scala/com/mesosphere/universe/bijection/UniverseConversions.scala]: Why are these two imported? Should we not be using the fully qualified name?

File: cosmos-bijection/src/test/scala/com/mesosphere/universe/bijection/TestUniverseConversions.scala [hunk 19]
```diff
@@ -2,10 +2,12 @@ package com.mesosphere.universe.bijection
 
 import com.mesosphere.universe
 import com.twitter.bijection.Conversion
+import com.twitter.bijection.Conversion.asMethod
+
```
Comment (larioj) [inline cosmos-bijection/src/test/scala/com/mesosphere/universe/bijection/TestUniverseConversions.scala]: useless space

File: cosmos-model/src/main/scala/com/mesosphere/universe/PackageDefinition.scala [hunk 20]
```diff
@@ -101,4 +101,20 @@ package v3.model {
     }
   }
 
+  sealed trait SupportedPackageDefinition
+    extends PackageDefinition
+
+  object SupportedPackageDefinition {
+
+    implicit val supportedPackageDefinitionOrdering = new Ordering[SupportedPackageDefinition] {
```
Comment (cruhland) [inline cosmos-model/src/main/scala/com/mesosphere/universe/PackageDefinition.scala]: You should be able to simplify this to:

```scala
implicit val supportedPackageDefinitionOrdering: Ordering[SupportedPackageDefinition] =
  Ordering.by(identity)
```

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/CosmosError.scala [hunk 21]
```diff
@@ -507,4 +507,10 @@ case class OperationInProgress(coordinate: rpc.v1.model.PackageCoordinate) exten
 case class InvalidPackage(reason: PackageUtil.PackageError) extends CosmosError(Some(reason)) {
   override val getData: Option[JsonObject] = Some(reason.getData)
 }
+
+case class UnknownMediaType(msg: String) extends CosmosError {
```
Comment (cruhland) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/CosmosError.scala]: I'm not sure that we need a `CosmosError` for this. If we ever encounter a media type that isn't expected, it indicates a bug in Cosmos; there was nothing the client could have done to prevent it. Throwing an `AssertionError` or a `RuntimeException` with an informative message should be enough in case we need to debug.

File: cosmos-model/src/main/scala/com/mesosphere/cosmos/rpc/v1/model/LocalPackage.scala [hunk 22]
```diff
@@ -2,7 +2,7 @@ package com.mesosphere.cosmos.rpc.v1.model
 
 import com.mesosphere.cosmos.storage.v1.model.Operation
 import com.mesosphere.universe
-import com.mesosphere.universe.v3.syntax.PackageDefinitionOps._
+import com.mesosphere.universe.v3.syntax.PackageDefinitionOps.packageDefinitionToPackageDefinitionOps
```
Comment (larioj) [inline cosmos-model/src/main/scala/com/mesosphere/cosmos/rpc/v1/model/LocalPackage.scala]: Change this back

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/storage/PackageStorage.scala [hunk 23]
```diff
@@ -22,21 +23,28 @@ final class PackageStorage private(objectStorage: ObjectStorage) {
...
+            new String(data, StandardCharsets.UTF_8)
+          )
```
Comment (cruhland) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/storage/PackageStorage.scala]: I think we discussed in standup changing this to the following?

```scala
case (universe.MediaTypes.universeV3Package, data) =>
  decode[universe.v3.model.V3Package](new String(data, StandardCharsets.UTF_8))
```

We can still return `Future[Option[universe.v3.model.SupportedPackageDefinition]]` since `V3Package` is a subclass.

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/storage/PackageStorage.scala (hunk 15 above)
Comment (cruhland) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/storage/PackageStorage.scala]: That's fine, since you have the default case, correct?

File: cosmos-model/src/main/scala/com/mesosphere/cosmos/storage/v1/model/PendingOperation.scala [hunk 24]
```diff
@@ -1,11 +1,11 @@
 package com.mesosphere.cosmos.storage.v1.model
 
 import com.mesosphere.cosmos.rpc
-import com.mesosphere.universe.v3.syntax.PackageDefinitionOps._
+import com.mesosphere.universe.v3.syntax.PackageDefinitionOps.packageDefinitionToPackageDefinitionOps
```
Comment (larioj) [inline cosmos-model/src/main/scala/com/mesosphere/cosmos/storage/v1/model/PendingOperation.scala]: Change this back.

File: cosmos-server/src/it/scala/com/mesosphere/cosmos/PackageAddSpec.scala [hunk 25]
```diff
@@ -19,6 +19,7 @@ import com.mesosphere.cosmos.test.TestUtil
 import com.mesosphere.universe
 import com.mesosphere.universe.test.TestingPackages
 import com.mesosphere.universe.v3.circe.Decoders._
+import com.mesosphere.universe.v3.model.SupportedPackageDefinition
```
Comment (larioj) [inline cosmos-server/src/it/scala/com/mesosphere/cosmos/PackageAddSpec.scala]: We should not need this import.

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/handler/PackageAddHandler.scala (hunk 16 above)
Comment (cruhland) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/handler/PackageAddHandler.scala]: We discussed this previously. This might require changing `EndpointHandler` to accommodate the new use case of `PackageAddHandler`.

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/storage/PackageStorage.scala (hunk 17 above)
Comment (cruhland) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/storage/PackageStorage.scala]: :+1:

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/repository/package.scala [hunk 26]
```diff
@@ -2,33 +2,34 @@ package com.mesosphere.cosmos
 
 import com.mesosphere.cosmos.storage.PackageStorage
 import com.mesosphere.universe
+import com.mesosphere.universe.v3.syntax.PackageDefinitionOps.packageDefinitionToPackageDefinitionOps
```
Comment (larioj) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/repository/package.scala]: Why do we need this import?

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/search/package.scala [hunk 27]
```diff
@@ -1,6 +1,7 @@
 package com.mesosphere.cosmos
 
 import cats.syntax.either._
+import com.mesosphere.universe.v3.syntax.PackageDefinitionOps.packageDefinitionToPackageDefinitionOps
```
Comment (larioj) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/search/package.scala]: We do not need this import

File: cosmos-model/src/main/scala/com/mesosphere/universe/PackageDefinition.scala (hunk 20 above)
Comment (larioj) [inline cosmos-model/src/main/scala/com/mesosphere/universe/PackageDefinition.scala]: I don't understand. What does ordering by identity accomplish.

File: cosmos-bijection/src/test/scala/com/mesosphere/universe/bijection/UniverseConversionsSpec.scala [hunk 28]
```diff
@@ -68,8 +68,10 @@ final class UniverseConversionsSpec extends FreeSpec with Matchers {
       val command = v3Package.command
 
       val roundTrip = v3Package
-        .as[(universe.v3.model.V3Metadata, universe.v3.model.ReleaseVersion)]
-        .as[universe.v3.model.V3Package]
+        .asInstanceOf[universe.v3.model.SupportedPackageDefinition]
+        .as[(universe.v3.model.Metadata, universe.v3.model.ReleaseVersion)]
+        .as[universe.v3.model.SupportedPackageDefinition]
+        .asInstanceOf[universe.v3.model.V3Package]
```
Comment (jsancio) [inline cosmos-bijection/src/test/scala/com/mesosphere/universe/bijection/UniverseConversionsSpec.scala]: What are you testing here? Is the description still correct `V3Package => ... => V3Package`?

File: cosmos-model/src/main/scala/com/mesosphere/universe/PackageDefinition.scala [hunk 29]
```diff
@@ -101,4 +101,20 @@ package v3.model {
...
+      override def compare(a: SupportedPackageDefinition, b: SupportedPackageDefinition): Int = {
+        PackageDefinition.compare(
+          (a.name, a.version, a.releaseVersion),
+          (b.name, b.version, b.releaseVersion)
+        )
+      }
+    }
+
+  }
```
Comment (jsancio) [inline cosmos-model/src/main/scala/com/mesosphere/universe/PackageDefinition.scala]: Let's move `SupportedPackageDefinition` after `PackageDefinition` and before `V2Package`.

File: cosmos-model/src/main/scala/com/mesosphere/universe/PackageDefinition.scala (hunk 20 above)
Comment (cruhland) [inline cosmos-model/src/main/scala/com/mesosphere/universe/PackageDefinition.scala]: It's maybe a bit cryptic if you don't know the definition of `Ordering.by`. It's delegating to the implicit `Ordering[PackageDefinition]` defined earlier in the file.

File: cosmos-bijection/src/main/scala/com/mesosphere/universe/bijection/UniverseConversions.scala [hunk 30]
```diff
@@ -97,7 +97,7 @@ object UniverseConversions {
       universe.v2.model.ReleaseVersion]
   }
 
-  implicit val v3MetadataToV3Package:
+  private implicit val v3MetadataToV3Package:
```
Comment (jsancio) [inline cosmos-bijection/src/main/scala/com/mesosphere/universe/bijection/UniverseConversions.scala]: `private[this]`.

File: cosmos-bijection/src/test/scala/com/mesosphere/universe/bijection/TestUniverseConversions.scala [hunk 31]
```diff
@@ -2,10 +2,11 @@ package com.mesosphere.universe.bijection
 
 import com.mesosphere.universe
 import com.twitter.bijection.Conversion
+import com.twitter.bijection.Conversion.asMethod
 
 object TestUniverseConversions {
 
-  implicit val v3PackageToV3Metadata:
+  private implicit val v3PackageToV3Metadata:
```
Comment (jsancio) [inline cosmos-bijection/src/test/scala/com/mesosphere/universe/bijection/TestUniverseConversions.scala]: `private[this]`

File: cosmos-model/src/main/scala/com/mesosphere/universe/PackageDefinition.scala (hunk 20 above)
Comment (larioj) [inline cosmos-model/src/main/scala/com/mesosphere/universe/PackageDefinition.scala]: Yup quite cryptic. But kinda makes sense.

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/handler/PackageAddHandler.scala (hunk 16 above)
Comment (jsancio) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/handler/PackageAddHandler.scala]: Yea. Didn't we agree to make this `case class AddResponse(package: SupportedPackageDefinition)` and change the jsonschema for `application/...add-response+json...` the union of all of the supported packages?

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/storage/PackageStorage.scala (hunk 15 above)
Comment (larioj) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/storage/PackageStorage.scala]: Yup

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/handler/PackageAddHandler.scala [hunk 32]
```diff
@@ -55,23 +55,26 @@ final class PackageAddHandler(
...
+          new rpc.v1.model.AddResponse(v3Package)
```
Comment (jsancio) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/handler/PackageAddHandler.scala]: Should `AddResponse` be a `case class` if so, no need for `new`.

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/storage/PackageStorage.scala (hunk 15 above)
Comment (jsancio) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/storage/PackageStorage.scala]: @larioj pattern matching doesn't match with the symbol is lowercase. I think you need to make it `UniverseV3Package` instead of `universeV3Package`.

Comment (larioj) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/storage/PackageStorage.scala]: Paths count as stable identifiers, and as such can be used as Literal pattern matches.

```scala
scala> import com.mesosphere.universe
import com.mesosphere.universe

scala> universe.MediaTypes.UniverseV3Repository match { case universe.MediaTypes.universeV3Package => true; }
scala.MatchError: MediaType(application,MediaTypeSubType(vnd.dcos.universe.repo,Some(json)),Map(charset -> utf-8, version -> v3)) (of class com.mesosphere.cosmos.http.MediaType)
  ... 43 elided
```

File: cosmos-bijection/src/test/scala/com/mesosphere/universe/bijection/UniverseConversionsSpec.scala (hunk 28 above)
Comment (larioj) [inline cosmos-bijection/src/test/scala/com/mesosphere/universe/bijection/UniverseConversionsSpec.scala]: Yes

Comment (jsancio) [inline cosmos-bijection/src/test/scala/com/mesosphere/universe/bijection/UniverseConversionsSpec.scala]: If that is the case then we should remove both `asInstanceOf`. This probably means that you will have to make the implicit converter public. If instead you are trying to test `SPD => ... SPD` then I think you just need to change the signature of this method to `(value: SPD) => ...`

File: cosmos-model/src/main/scala/com/mesosphere/universe/PackageDefinition.scala (hunk 20 above)
Comment (larioj) [inline cosmos-model/src/main/scala/com/mesosphere/universe/PackageDefinition.scala]: For some reason this broke the tests.

Comment (cruhland) [inline cosmos-model/src/main/scala/com/mesosphere/universe/PackageDefinition.scala]: This should work:

```scala
    implicit val supportedPackageDefinitionOrdering: Ordering[SupportedPackageDefinition] =
      PackageDefinition.packageDefinitionOrdering.on(identity)
```

File: cosmos-model/src/main/scala/com/mesosphere/universe/PackageDefinition.scala (hunk 29 above)
Comment (larioj) [inline cosmos-model/src/main/scala/com/mesosphere/universe/PackageDefinition.scala]: Done

File: cosmos-server/src/main/scala/com/mesosphere/cosmos/handler/PackageAddHandler.scala (hunk 16 above)
Comment (larioj) [inline cosmos-server/src/main/scala/com/mesosphere/cosmos/handler/PackageAddHandler.scala]: We'll make that change when we add V4Package
